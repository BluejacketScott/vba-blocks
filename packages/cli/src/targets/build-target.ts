import { join } from 'path';
import { pathExists } from 'fs-extra';
import { Config } from '../config';
import { Target, Source, Reference } from '../manifest';
import { importManifest } from '../addin';

export default async function buildTarget(
  config: Config,
  target: Target,
  graph: { src: Source[]; references: Reference[] }
) {
  const file = join(config.build, `${target.name}.${target.type}`);
  if (!await pathExists(file)) {
    throw new Error(
      `Target binary for ${target.name}.${target.type} not found`
    );
  }

  await importManifest(config, target, graph);
}
