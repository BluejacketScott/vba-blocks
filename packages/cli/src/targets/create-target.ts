import { join } from 'path';
import { promisify } from 'util';
import { pathExists, ensureDir, remove } from 'fs-extra';
import { Config } from '../config';
import { Target } from '../manifest';
import { zip } from '../utils';
const copyFile = promisify(require('fs').copyFile);

export default async function createTarget(
  config: Config,
  target: Target
): Promise<string> {
  const dir = join(config.cwd, target.path);
  const file = join(config.build, `${target.name}.${target.type}`);
  const backup = join(config.build, '.backup', `${target.name}.${target.type}`);

  if (!await pathExists(dir)) {
    throw new Error(
      `Path "${target.path}" not found for target "${target.name}"`
    );
  }

  // Backup file if target already exists
  if (await pathExists(backup)) await remove(backup);
  if (await pathExists(file)) {
    await copyFile(file, backup);
    await remove(file);
  }

  // Zip directory to create target
  await ensureDir(config.build);
  await zip(dir, file);

  return file;
}
