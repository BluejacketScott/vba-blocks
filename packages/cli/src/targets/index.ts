import { unixJoin } from '../utils';
import { Target } from '../manifest';
import { Project } from '../project';

export { BuildGraph, createBuildGraph } from './build-graph';
export { default as stageBuildGraph } from './stage-build-graph';
export { default as createTarget } from './create-target';
export { default as buildTarget } from './build-target';

export function getFile(project: Project, target: Target): string {
  return unixJoin(project.paths.build, `${target.name}.${target.type}`);
}

export function getBackup(project: Project, target: Target): string {
  return unixJoin(project.paths.backup, `${target.name}.${target.type}`);
}
