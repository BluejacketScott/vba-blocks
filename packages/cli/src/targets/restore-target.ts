import { Project } from '../project';
import { Target } from '../manifest';
import { getBackup, getFile } from './';
import { pathExists, copyFile } from '../utils';
import { targetRestoreFailed } from '../errors';

export default async function restoreTarget(project: Project, target: Target) {
  const backup = getBackup(project, target);
  const file = getFile(project, target);

  if (!await pathExists(backup)) return;

  try {
    await copyFile(backup, file);
  } catch (err) {
    throw targetRestoreFailed(backup, file, err);
  }
}
