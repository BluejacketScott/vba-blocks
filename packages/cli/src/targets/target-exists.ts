import { Config } from '../config';
import { Target } from '../manifest';

export default async function targetExists(
  config: Config,
  target: Target
): Promise<boolean> {
  // TODO
  return false;
}
