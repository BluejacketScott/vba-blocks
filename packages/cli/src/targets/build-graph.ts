import { Source, Reference } from '../manifest';
import { Project, fetchDependencies, loadManifests } from '../project';
import { BuildOptions } from './build-target';

export interface BuildGraph {
  src: Source[];
  references: Reference[];
}

export async function createBuildGraph(
  project: Project,
  options: BuildOptions
): Promise<BuildGraph> {
  const src: Map<string, Source> = new Map();
  const references: Map<string, Reference> = new Map();

  const manifests = project.manifests || (await loadManifests(project));

  for (const manifest of manifests) {
    for (const source of manifest.src) {
      const { name } = source;
      if (src.has(name))
        throw new Error(`Conflicting source named "${source.name}"`);
      src.set(name, source);
    }
    for (const reference of manifest.references) {
      const { name, guid } = reference;
      if (references.has(name) && references.get(name)!.guid !== guid)
        throw new Error(`Conficting reference named "${reference.name}"`);
      references.set(name, reference);
    }
  }

  return { src: [...src.values()], references: [...references.values()] };
}
