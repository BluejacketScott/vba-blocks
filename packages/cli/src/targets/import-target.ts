import { Project } from '../project';
import { Target } from '../manifest';
import { importGraph } from '../addin';
import { BuildGraph } from './build-graph';
import stageBuildGraph from './stage-build-graph';

export default async function importTarget(
  project: Project,
  target: Target,
  graph: BuildGraph
) {
  graph = await stageBuildGraph(graph);
  await importGraph(project, target, graph);
}
