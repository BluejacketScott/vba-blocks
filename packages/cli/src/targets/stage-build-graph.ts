import { basename } from 'path';
import { BuildGraph } from './build-graph';
import { getStaging, ensureDir, copyFile, parallel, unixJoin } from '../utils';
import env from '../env';

// To avoid "Grant File Access" prompts on Mac,
// files to be imported/exported from VBA are staged to
// a location that can freely share files with Office:
// "HOME/Library/Group Containers/##########.Office"
//
// See: http://www.rondebruin.nl/mac/mac034.htm

export default async function stageBuildGraph(
  graph: BuildGraph
): Promise<BuildGraph> {
  if (env.isWindows) return graph;

  const staging = env.staging || (env.staging = await getStaging());

  const src = await parallel(
    graph.src,
    async source => {
      const { name, path, optional } = source;

      const dest = unixJoin(staging, basename(source.path));
      await copyFile(source.path, dest);

      return { name, path: dest, optional, original: path };
    },
    { progress: env.reporter.progress('Staging dependencies') }
  );

  return Object.assign({}, graph, { src });
}
