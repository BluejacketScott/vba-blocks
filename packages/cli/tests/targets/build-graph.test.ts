import { createBuildGraph } from '../../src/targets/build-graph';

test('should create build graph', async () => {});
