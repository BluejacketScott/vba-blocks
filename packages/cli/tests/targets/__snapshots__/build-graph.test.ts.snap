// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create build graph 1`] = `
Object {
  "references": Array [],
  "src": Array [
    Object {
      "name": "a",
      "optional": false,
      "path": "C:\\\\dev\\\\vba-blocks\\\\packages\\\\cli\\\\tests\\\\fixtures\\\\sources\\\\a\\\\a.bas",
    },
    Object {
      "name": "b",
      "optional": false,
      "path": "C:\\\\dev\\\\vba-blocks\\\\packages\\\\cli\\\\tests\\\\fixtures\\\\sources\\\\b\\\\b.cls",
    },
    Object {
      "name": "d",
      "optional": false,
      "path": "C:\\\\dev\\\\vba-blocks\\\\packages\\\\cli\\\\tests\\\\fixtures\\\\sources\\\\d\\\\d.cls",
    },
    Object {
      "name": "f",
      "optional": false,
      "path": "C:\\\\dev\\\\vba-blocks\\\\packages\\\\cli\\\\tests\\\\fixtures\\\\sources\\\\f\\\\f.bas",
    },
    Object {
      "name": "g",
      "optional": false,
      "path": "C:\\\\dev\\\\vba-blocks\\\\packages\\\\cli\\\\tests\\\\fixtures\\\\sources\\\\g\\\\g.cls",
    },
    Object {
      "name": "c",
      "optional": false,
      "path": "C:\\\\dev\\\\vba-blocks\\\\packages\\\\cli\\\\tests\\\\fixtures\\\\sources\\\\c\\\\c.bas",
    },
  ],
}
`;
